x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java

```diff
@@ -463,6 +451,60 @@ public void testInvalidateApiKeyWillClearApiKeyCache() throws IOException, Execu
         assertThat(e.getResponse().getStatusLine().getStatusCode(), is(RestStatus.UNAUTHORIZED.getStatus()));
     }
 
+    public void testDynamicDeletionInterval() throws Exception {
+        try {
+            // Set retention period to be 1 ms, and delete interval to be 1 hour
+            long deleteIntervalMs = 3600_000;
+            Settings.Builder builder = Settings.builder();
+            builder.put(ApiKeyService.DELETE_RETENTION_PERIOD.getKey(), TimeValue.timeValueMillis(1));
+            builder.put(ApiKeyService.DELETE_INTERVAL.getKey(), TimeValue.timeValueMillis(deleteIntervalMs));
+            updateClusterSettings(builder);
+
+            // Create API keys to test
+            List<CreateApiKeyResponse> responses = createApiKeys(3, null).v1();
+            String[] apiKeyIds = responses.stream().map(CreateApiKeyResponse::getId).toArray(String[]::new);
+
+            // Invalidate one API key to trigger run of inactive remover (will run once and then after DELETE_INTERVAL)
+            PlainActionFuture<InvalidateApiKeyResponse> listener = new PlainActionFuture<>();
+            Client client = authorizedClient();
+            client.execute(InvalidateApiKeyAction.INSTANCE, InvalidateApiKeyRequest.usingApiKeyId(apiKeyIds[0], false), listener);
+            verifyInvalidateResponse(1, Collections.singletonList(responses.get(0)), listener.get());
+            awaitApiKeysRemoverCompletion();
+
+            // Get API keys to make sure remover didn't remove any yet
+            assertThat(getAllApiKeyInfo(client, false).length, equalTo(3));
+
+            // Invalidate another key
+            listener = new PlainActionFuture<>();
+            client.execute(InvalidateApiKeyAction.INSTANCE, InvalidateApiKeyRequest.usingApiKeyId(apiKeyIds[1], false), listener);
+            verifyInvalidateResponse(1, Collections.singletonList(responses.get(1)), listener.get());
+            awaitApiKeysRemoverCompletion();
+
+            // Get API keys to make sure remover didn't remove any yet (shouldn't be removed because of the long DELETE_INTERVAL)
+            assertThat(getAllApiKeyInfo(client, false).length, equalTo(3));
+
+            // Update DELETE_INTERVAL to every 1 ms
+            builder = Settings.builder();
+            deleteIntervalMs = 1;
```

Comment:
Come to think of it: could be `0` to avoid the (extremely unlikely) case that the invalidation call happens in under 1ms after the call here.

Reply:
Yes, that would probably be fine, but even if we set it to `0` there is no guarantee that the remover will run before the invalidation, even though it probably always will. I think doing the "safe assertion" you suggested in your other comment makes the most sense.